src/symbol.c

```diff
@@ -0,0 +1,167 @@
+#define R_NO_REMAP
+#include <R.h>
+#include <Rinternals.h>
+#include <R_ext/GraphicsEngine.h>
+#include <stdbool.h>
+
+#define attribute_hidden
+
+// Interface functions ---------------------------------------------------------
+
+void copy_character(SEXP tgt, SEXP src, R_xlen_t len);
+R_xlen_t unescape_character_in_copy(SEXP tgt, SEXP src, R_xlen_t i);
+SEXP unescape_sexp(SEXP name);
+
+SEXP rlang_symbol(SEXP chr) {
+  SEXP string = STRING_ELT(chr, 0);
+  return Rf_install(Rf_translateChar(string));
+}
+
+SEXP rlang_symbol_to_character(SEXP chr) {
+  SEXP name = PRINTNAME(chr);
+  return Rf_ScalarString(unescape_sexp(name));
+}
+
+SEXP rlang_unescape_character(SEXP chr) {
+  R_xlen_t len = Rf_xlength(chr);
+  R_xlen_t i = unescape_character_in_copy(R_NilValue, chr, 0);
+  if (i == len) return chr;
+
+  SEXP ret = PROTECT(Rf_allocVector(STRSXP, len));
+  copy_character(ret, chr, i);
+  unescape_character_in_copy(ret, chr, i);
+  UNPROTECT(1);
+  return ret;
+}
+
+// Private functions -----------------------------------------------------------
+
+bool has_unicode_escape(const char* chr);
+int unescape_char(char* chr);
+int unescape_char_found(char* chr);
+int process_byte(char* tgt, char* const src, int* len_processed);
+bool has_codepoint(const char* src);
+bool is_hex(const char chr);
+
+void copy_character(SEXP tgt, SEXP src, R_xlen_t len) {
```

Comment:
hmm yeah but we might convert those to C++ (for enum templating) and we might ship them as header-only library in the future. Don't worry about it, we can keep `copy_character()` in symbol.c and figure this out later.